zonemaster/zonemaster-backend
# Harmonize databases

## Purpose

This PR tries to improve and add consistency between database engines. There are PRs to update the database in a way that breaks things (new `undelegated` column in PosgreSQL for instance). It might be nice to take the opportunity of this new Backend release to improve and fix some behaviors in the databases.

A first part in this PR is to improve the `create_db()` routine in order to be able to call it at daemon startup (which will be handle in another PR #846 ).

## Context

Based on top of #833 and #834
Addresses #832, #840

## Changes

There are several changes. This PR can easily be split in smaller PRs.

* (1/11) Create indexes in SQLite -> #865

* (2/11) Always check that a table, index or trigger does not exists before creation. For some cases, the clause `IF NOT EXISTS` is sufficient, for others, we need workaround because it is not present in all supported versions of the database engine. (069732e) -> #866 

* (3/11) Rename column `params_deterministic_hash` into `fingerprint` (26c0511)

* (4/11) Correctly use `creation_time` in SQLite to check the age of a test (a8a4cee)

* (5/11) Set `test_start_time` to NULL by default and when the test is rescheduled (6aafdeb)

* (6/11) Move the `schedule_for_retry` routine to "DB.pm" (it now uses a SQL syntax that is valid for all database engines) (29e7eb2)

* (7/11) Fix a nonexistent index key field in PosgreSQL (7145d4f)

* (8/11) Add a new `domain` column in PostgreSQL to be consistent with other database engines (4e4d563)

* (9/11) Update the "users" table: (8563602)
   * new `username` and `api_key` columns for PostgreSQL
   * remove the `user_info` column for SQLite, MySQL and PostgreSQL

* (10/11) Made some refactoring: (576dc86)
    * New method `recent_test_hash_id' to search for pre-existing test
    * Move `create_new_batch_job' method to DB.pm
    * Move `get_test_params' method to DB.pm
    * Use generic DBI `last_insert_id'

* (11/11) Compute the `hash_id` before its insertion in database in order to use the same `create_new_test` method across database engines (60da83a)

* For all the changes that affect the database, each patch script has been updated accordingly
 
 

## How to test this PR

Do one of the following:
* upgrade an existing database with the corresponding patch script
* install and initialize the database with the instructions in the [installation document](https://github.com/zonemaster/zonemaster-backend/blob/develop/docs/Installation.md)

Then check for the following things:
* SQLite `test_results` table has indexes, `PRAGMA index_info(test_results)` should result in 5 indexes
* there is no `params_deterministic_hash` column, it should be named `fingerprint`
* there is no index `test_results__params_deterministic_hash`, it should be named `test_results__fingerprint`
* for PostgreSQL, a column `domain` should exist in the `test_results` table (if migrating, the field should be properly initialized)
* for PostgreSQL, the columns `username` and `api_key` should exist in the `users` table (if migrating, the fields should be properly initialized)
* the `hash_id` field can't be null and has no default value
* there should not be any `user_info` field in the `users` table anymore

Finally run some checks, you should be able to properly use the following methods:
* `start_domain_test` (and you should be able to reuse a pre-existing test)
* `add_api_user`
* `add_batch_job`
* `get_test_params`
